cmake/add_enclave.cmake

```diff
@@ -101,3 +147,91 @@ function(add_enclave)
   # Add a custom target with `ALL` semantics so these targets are always built.
   add_custom_target(${ENCLAVE_TARGET}_signed_target ALL DEPENDS ${SIGNED_LOCATION})
 endfunction()
+
+macro(add_enclave_optee)
+   set(options CXX)
+   set(oneValueArgs TARGET UUID KEY)
+   set(multiValueArgs SOURCES)
+   cmake_parse_arguments(ENCLAVE
+     "${options}"
+     "${oneValueArgs}"
+     "${multiValueArgs}"
+     ${ARGN})
+
+  # Set up the linker flags exactly as we need them such that the resulting
+  # binary be compatible with OP-TEE's loader.
+  set(CMAKE_SHARED_LIBRARY_LINK_C_FLAGS)
+  set(CMAKE_SHARED_LIBRARY_LINK_CXX_FLAGS)
+  set(CMAKE_EXE_EXPORTS_C_FLAG)
+
+  string(REPLACE "gcc" "ld" LINKER ${CMAKE_C_COMPILER})
+  set(CMAKE_C_LINK_EXECUTABLE "${LINKER} <CMAKE_CXX_LINK_FLAGS> <LINK_FLAGS> <OBJECTS> <LINK_LIBRARIES> -o <TARGET>")
+  set(CMAKE_CXX_LINK_EXECUTABLE "${LINKER} <CMAKE_CXX_LINK_FLAGS> <LINK_FLAGS> <OBJECTS> <LINK_LIBRARIES> -o <TARGET>")
+
+  # Generate linker script from template.
+  string(REPLACE "gcc" "cpp" C_PREPROCESSOR ${CMAKE_C_COMPILER})
+  set(TA_LINKER_SCRIPT ${CMAKE_CURRENT_BINARY_DIR}/ta.ld)
+  set(TA_LINKER_SCRIPT ${TA_LINKER_SCRIPT} PARENT_SCOPE)
+  add_custom_target(${ENCLAVE_TARGET}.ld
```

Comment:
Should their be a `DEPENDS` for this (and the other custom targets)? I suspect `SOURCES` was used with the intention of setting up a dependency, but thanks to CMake being CMake, this actually only has the side effect of grouping the files for IDEs parsing CMake projects. `DEPENDS` is what will actually set the dependencies.

Reply:
Good catch. The other two custom targets are followed by a call to `add_dependencies`, which CMake's documentation suggests should be used for custom targets to depend on other CMake targets.